File: install/ignore/sdk/ruby/lib/airway-im-sdk-ruby/message.rb
# frozen_string_literal: true

module AirwayIM
  # A message response that behaves like a Hash (JSON serialization,
  # equality, [] access) while also exposing its fields as reader methods:
  # message.id, message.sequence, message.sender.uuid, and so on. A Hash
  # subclass on purpose so both access styles stay equivalent.
  class Message < Hash
    FIELDS = %w[id conversation_id content content_type created_at sequence].freeze

    def initialize(data = {})
      super()
      update(data || {})
    end

    FIELDS.each do |field|
      define_method(field) { self[field] }
    end

    def sender
      Sender.new(self["sender"] || {})
    end
  end

  # Author profile inside a Message, same Hash-with-readers contract:
  # message.sender.uuid or message["sender"]["uuid"].
  class Sender < Hash
    FIELDS = %w[uuid username nickname avatar_url].freeze

    def initialize(data = {})
      super()
      update(data || {})
    end

    FIELDS.each do |field|
      define_method(field) { self[field] }
    end
  end
end

File: install/ignore/sdk/php/src/Message.php
<?php

declare(strict_types=1);

namespace AirwayIM;

/**
 * A message response that behaves like a JSON object (array access, JSON
 * serialization) while also exposing its fields as property reads:
 * $message->id, $message->sequence, $message->sender->uuid, and so on.
 */
final class Message extends DataObject
{
    public function __get(string $name)
    {
        if ($name === 'sender') {
            $sender = $this->data['sender'] ?? null;
            return new Sender(is_array($sender) ? $sender : []);
        }
        return parent::__get($name);
    }
}
